data/worlds/README.md

```diff
@@ -0,0 +1,189 @@
+# World DSL guide
+
+Swarm has a built-in, special-purpose *domain specific language* for
+describing worlds (*i.e.* the terrain, entities, other robots,
+etc. which the player sees on the map when a scenario is loaded).  It
+is somewhat bare bones at the moment, but will continue to develop.
+
+## Overview
+
+The basic idea of the world DSL is ultimately to describe a *function*
+which specifies a *cell* value for every coordinate.  In addition,
+this is done in such a way that all randomness used for procedural
+generation (if any) is ultimately derived from a single seed, so world
+generation is actually 100% deterministic and reproducible by design.
+
+## Types
+
+- There are four base types: `bool`, `int`, `float`, and
+  `cell`.
+  - `bool` values are written `true` and `false`.
+  - `int` values are written like `3` or `-5`.
+  - `float` values are written like `3.2` or `0.0`.  Note that `0` is
+    always an `int`, and `0.0` is always a `float`.   If you use `0`
+    as an argument to a function expecting a `float`, it is a type
+    error!  This may be slightly annoying but it keeps typechecking
+    much simpler.
+  - `cell` values describe the contents of a world cell (terrain,
+    entity, etc.), and are explained in more detail below.
+
+- In addition, if `b` is a base type, `World b` is a type representing a
+  "world function" which specifies a value of type `b` at every
+  coordinate in its domain.
+
+- Any base type `b` is a subtype of `World b`; that is, a value of
+  type `b` may be used anywhere a `World b` is expected, or, put
+  another way, any `b` may be automatically "promoted" to a `World b`.
+  Semantically, a single value of a base type may be promoted to an
+  infinite, constant world function which returns that single value at
+  every coordinate.
+
+- In general, any function of type `b1 -> ... -> bn` where the `bi`
+  are all base types may be "lifted" to have type `World b1 -> ... ->
+  World bn`, which means the function will act "coordinatewise",
+  i.e. like a giant 2D `zipWith`.
+
+    - For example, the `<` operator has a type like `integer ->
+      integer -> bool` but that means it could also have type `World
+      integer -> World integer -> World bool`.
+
+## Syntax
+
+Comments are specified by `//` (single line) or `/* ... */`
+(multi-line).
+
+Identifiers consist of any non-empty sequence of letters, digits,
+underscores, or single quotes, but must begin with a letter or
+underscore.
+
+The extended BNF syntax `S*,` denotes the syntax `S` repeated zero
+or more times, separated by `,`.  Likewise, `S+,` denotes one or more
+repetitions of `S` separated by `,`.
+
+```
+<int> ::= integer literal
```

Comment:
I like the formal specification here, could we check that it applies to all World DSL?

I expect we use megaparsec.

Reply:
Or other way - create this description from code?

Reply:
Not easily at the moment, since it's not fully formal (notice it contains English phrases like "any single character besides...".  In theory we could have a fully formal specification from which we extract both the parser code and some documentation, but it sounds like a lot of work.  I'll just add a note in the code as a reminder to update the README to match when modifying the parser.